### PR TITLE
do not identify function types by throwing exceptions

### PR DESCRIPTION
**Pull request by richardstartin**
This PR allows function (transform/aggregation) type to be identified without throwing exceptions when an expression (e.g. "ASC") is not a function type, this reduces allocation rate.

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionDefinitionRegistry.java [hunk 1]
```diff
@@ -29,20 +29,10 @@ private FunctionDefinitionRegistry() {
   }
 
   public static boolean isAggFunc(String functionName) {
```
Comment (mqliang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/FunctionDefinitionRegistry.java:31]: This function looks like just an alias of `AggregationFunctionType.isAggregationFunction`.  Can we just remove it and let caller to call `AggregationFunctionType.isAggregationFunction` directly?

File: pinot-common/src/main/java/org/apache/pinot/common/function/FunctionDefinitionRegistry.java [hunk 2]
```diff
@@ -29,20 +29,10 @@ private FunctionDefinitionRegistry() {
...
   }
 
   public static boolean isTransformFunc(String functionName) {
-    try {
-      TransformFunctionType.getTransformFunctionType(functionName);
-      return true;
-    } catch (IllegalArgumentException e) {
-      return false;
-    }
+    return TransformFunctionType.isTransformFunction(functionName);
```
Comment (mqliang) [inline pinot-common/src/main/java/org/apache/pinot/common/function/FunctionDefinitionRegistry.java:36]: ditto

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java [hunk 3]
```diff
@@ -81,14 +87,28 @@ public String getName() {
...
+
   /**
    * Returns the corresponding aggregation function type for the given function name.
    * <p>NOTE: Underscores in the function name are ignored.
    */
   public static AggregationFunctionType getAggregationFunctionType(String functionName) {
-    String upperCaseFunctionName = StringUtils.remove(functionName, '_').toUpperCase();
-    if (upperCaseFunctionName.startsWith("PERCENTILE")) {
-      String remainingFunctionName = upperCaseFunctionName.substring(10);
+    if (functionName.regionMatches(true, 0, "percentile", 0, 10)) {
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java:110]: Does this give better performance than `startsWith()`? In most cases, the query won't match this if check

**Review by Jackie-Jiang — COMMENTED** [review]
Agree on avoiding the exception in normal execution path. Do we have some performance number back this change?

**Comment by richardstartin** [comment]
> Agree on avoiding the exception in normal execution path. Do we have some performance number back this change?

I had numbers for a few of these PRs (which have been submitted separately because some users prefer tightly scoped changes) and this was just an incremental improvement. I can measure this separately.

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java [hunk 4]
```diff
@@ -101,12 +107,28 @@
...
 
+  public static boolean isTransformFunction(String functionName) {
+    if (NAMES.contains(functionName)) {
+      return true;
+    }
+    // scalar functions
+    if (FunctionRegistry.containsFunction(functionName)) {
+      return true;
+    }
+    return NAMES.contains(functionName.toUpperCase().replace("_", ""));
```
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:129]: I honestly have no idea, but it's very hard to detect fast/slow code by eye.

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java (hunk 3 above)
Comment (richardstartin) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java:110]: This is case insensitive without allocation.

Comment (richardstartin) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java:110]: `regionMatches` feels like the better option to me:

```
Benchmark                                              (_function)  Mode  Cnt     Score     Error   Units
StartsWith.regionMatches                            percentile_foo  avgt    5    14.793 ±   0.467   ns/op
StartsWith.regionMatches:·gc.alloc.rate             percentile_foo  avgt    5    ≈ 10⁻⁴            MB/sec
StartsWith.regionMatches:·gc.alloc.rate.norm        percentile_foo  avgt    5    ≈ 10⁻⁵              B/op
StartsWith.regionMatches:·gc.count                  percentile_foo  avgt    5       ≈ 0            counts
StartsWith.regionMatches                            PERCENTILE_FOO  avgt    5    26.785 ±   0.480   ns/op
StartsWith.regionMatches:·gc.alloc.rate             PERCENTILE_FOO  avgt    5    ≈ 10⁻⁴            MB/sec
StartsWith.regionMatches:·gc.alloc.rate.norm        PERCENTILE_FOO  avgt    5    ≈ 10⁻⁵              B/op
StartsWith.regionMatches:·gc.count                  PERCENTILE_FOO  avgt    5       ≈ 0            counts
StartsWith.regionMatches                             doesn't match  avgt    5     7.091 ±   0.034   ns/op
StartsWith.regionMatches:·gc.alloc.rate              doesn't match  avgt    5    ≈ 10⁻⁴            MB/sec
StartsWith.regionMatches:·gc.alloc.rate.norm         doesn't match  avgt    5    ≈ 10⁻⁶              B/op
StartsWith.regionMatches:·gc.count                   doesn't match  avgt    5       ≈ 0            counts
StartsWith.startsWith                               percentile_foo  avgt    5    36.754 ±   0.447   ns/op
StartsWith.startsWith:·gc.alloc.rate                percentile_foo  avgt    5   965.883 ±  11.124  MB/sec
StartsWith.startsWith:·gc.alloc.rate.norm           percentile_foo  avgt    5    56.000 ±   0.001    B/op
StartsWith.startsWith:·gc.churn.G1_Eden_Space       percentile_foo  avgt    5   970.448 ±   4.635  MB/sec
StartsWith.startsWith:·gc.churn.G1_Eden_Space.norm  percentile_foo  avgt    5    56.265 ±   0.591    B/op
StartsWith.startsWith:·gc.churn.G1_Old_Gen          percentile_foo  avgt    5     0.001 ±   0.005  MB/sec
StartsWith.startsWith:·gc.churn.G1_Old_Gen.norm     percentile_foo  avgt    5    ≈ 10⁻⁴              B/op
StartsWith.startsWith:·gc.count                     percentile_foo  avgt    5    40.000            counts
StartsWith.startsWith:·gc.time                      percentile_foo  avgt    5    22.000                ms
StartsWith.startsWith                               PERCENTILE_FOO  avgt    5    16.395 ±   0.523   ns/op
StartsWith.startsWith:·gc.alloc.rate                PERCENTILE_FOO  avgt    5    ≈ 10⁻⁴            MB/sec
StartsWith.startsWith:·gc.alloc.rate.norm           PERCENTILE_FOO  avgt    5    ≈ 10⁻⁵              B/op
StartsWith.startsWith:·gc.count                     PERCENTILE_FOO  avgt    5       ≈ 0            counts
StartsWith.startsWith                                doesn't match  avgt    5    31.228 ±   1.392   ns/op
StartsWith.startsWith:·gc.alloc.rate                 doesn't match  avgt    5  1137.484 ±  48.866  MB/sec
StartsWith.startsWith:·gc.alloc.rate.norm            doesn't match  avgt    5    56.000 ±   0.001    B/op
StartsWith.startsWith:·gc.churn.G1_Eden_Space        doesn't match  avgt    5  1139.195 ± 256.337  MB/sec
StartsWith.startsWith:·gc.churn.G1_Eden_Space.norm   doesn't match  avgt    5    56.079 ±  11.876    B/op
StartsWith.startsWith:·gc.churn.G1_Old_Gen           doesn't match  avgt    5     0.002 ±   0.005  MB/sec
StartsWith.startsWith:·gc.churn.G1_Old_Gen.norm      doesn't match  avgt    5    ≈ 10⁻⁴              B/op
StartsWith.startsWith:·gc.count                      doesn't match  avgt    5    47.000            counts
StartsWith.startsWith:·gc.time                       doesn't match  avgt    5    25.000                ms
```

File: pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (richardstartin) [reply pinot-common/src/main/java/org/apache/pinot/common/function/TransformFunctionType.java:129]: It looks like `StringUtils.remove` is quite a lot faster on JDK11 when there are characters to remove, but allocates a little bit more, I will change this back.

**Comment by richardstartin** [comment]
@Jackie-Jiang here are some numbers for the query I was tuning (on my MacBook so incomparable to numbers on different PRs)

master
```
Benchmark                                                                               (_intBaseValue)  (_numRows)  Mode  Cnt        Score          Error   Units
BenchmarkFilteredAggregations.testNonFilteredAggregations                                             0     1500000  avgt    5    32506.601 ±     4192.330   us/op
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.alloc.rate                              0     1500000  avgt    5       22.952 ±        2.747  MB/sec
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.alloc.rate.norm                         0     1500000  avgt    5  1171670.000 ±    15616.821    B/op
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.churn.G1_Eden_Space                     0     1500000  avgt    5       37.526 ±      323.107  MB/sec
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.churn.G1_Eden_Space.norm                0     1500000  avgt    5  1914496.826 ± 16484395.047    B/op
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.churn.G1_Old_Gen                        0     1500000  avgt    5        7.749 ±       66.725  MB/sec
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.churn.G1_Old_Gen.norm                   0     1500000  avgt    5   395366.090 ±  3404221.273    B/op
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.count                                   0     1500000  avgt    5        1.000                 counts
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.time                                    0     1500000  avgt    5       13.000                     ms
```

branch
```
Benchmark                                                                      (_intBaseValue)  (_numRows)  Mode  Cnt       Score       Error   Units
BenchmarkFilteredAggregations.testNonFilteredAggregations                                    0     1500000  avgt    5   28881.564 ±  2888.988   us/op
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.alloc.rate                     0     1500000  avgt    5      20.376 ±     1.846  MB/sec
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.alloc.rate.norm                0     1500000  avgt    5  922164.440 ± 30175.150    B/op
BenchmarkFilteredAggregations.testNonFilteredAggregations:·gc.count                          0     1500000  avgt    5         ≈ 0              counts
```

**Comment by richardstartin** [comment]
The test failure should be addressed by #8154

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java [hunk 5]
```diff
@@ -81,14 +88,28 @@ public String getName() {
...
 
+  public static boolean isAggregationFunction(String functionName) {
+    if (NAMES.contains(functionName)) {
+      return true;
+    }
+    if (functionName.regionMatches(true, 0, "percentile", 0, 10)) {
+      try {
+        getAggregationFunctionType(functionName);
+        return true;
+      } catch (Exception ignore) {
```
Comment (Jackie-Jiang) [inline pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java:99]: We can directly return `false` here?

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java (hunk 3 above)
Comment (Jackie-Jiang) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java:111]: I believe the allocation is from the `StringUtils.remove(functionName, '_').toUpperCase()` which cannot be avoided in most of cases (non-percentile function).
A little behavior change here if we use `regionMatches` is that `percent_ile` cannot be recognized (not really important though).
Either way is fine, you can make the decision

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java (hunk 5 above)
Comment (richardstartin) [reply pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/AggregationFunctionType.java:99]: agreed

**Pull request merged**